toolkit/tools/graphscrubber/graphscrubber.go

```diff
@@ -0,0 +1,159 @@
+
+				logger.Log.Debugf("Adding a 'PreBuilt' node '%s' with id %d. For '%s'", preBuiltNode.FriendlyName(), preBuiltNode.ID(), parentNode.FriendlyName())
+				err = pkgGraph.AddEdge(parentNode, preBuiltNode)
+
+				if err != nil {
+					logger.Log.Errorf("Adding edge failed for %v -> %v", parentNode, preBuiltNode)
+				}
+			}
+		}
+	}
+
+	return
+}
+
+func removeExistingGoalNode(pkgGraph *pkggraph.PkgGraph, goalNodeName string) {
+	goalNode := pkgGraph.FindGoalNode(goalNodeName)
+	if goalNode != nil {
+		logger.Log.Infof("Found goalNode: %s. Removing it.", goalNodeName)
+		pkgGraph.RemoveNode(goalNode.ID())
+	} else {
+		logger.Log.Warnf("Can't find goalNode: %s. Skipping Removal", goalNodeName)
+	}
+	return
+}
+
+func subGraphPkgsToBuild(pkgGraph *pkggraph.PkgGraph, packageVersToBuild []*pkgjson.PackageVer) (optimizedGraph *pkggraph.PkgGraph, err error) {
+	optimizedGraph = pkgGraph
+
+	pkgToBuildGoalNode := pkgGraph.FindGoalNode(allGoalNodeName)
+	if schedulerutils.CanSubGraph(pkgGraph, pkgToBuildGoalNode, true) {
+		optimizedGraph, err = pkgGraph.CreateSubGraph(pkgToBuildGoalNode)
+		if err != nil {
+			logger.Log.Panicf("Failed to create subgraph error: %s", err)
+			return
+		}
+	} else {
+		logger.Log.Warnf("Not possible to subgraph")
+		return
+	}
+
+	return
+}
+
+func main() {
+	app.Version(exe.ToolkitVersion)
+	kingpin.MustParse(app.Parse(os.Args[1:]))
+	logger.InitBestEffort(*logFile, *logLevel)
+
+	scrubbedGraph := pkggraph.NewPkgGraph()
+
+	err := pkggraph.ReadDOTGraphFile(scrubbedGraph, *inputGraphFile)
+	if err != nil {
+		logger.Log.Panicf("Failed to read graph to file, %s. Error: %s", *inputGraphFile, err)
+	}
+
+	if !*hydratedBuild {
+		err = pkggraph.WriteDOTGraphFile(scrubbedGraph, *outputGraphFile)
+		if err != nil {
+			logger.Log.Panicf("Failed to write cache graph to file, %s. Error: %s", *outputGraphFile, err)
+		}
+		return
+	}
+
+	// Generate the list of packages that need to be built.
+	// If none are requested then all packages will be built.
+	packagesNamesToBuild := exe.ParseListArgument(*pkgsToBuild)
+	packagesNamesToRebuild := exe.ParseListArgument(*pkgsToRebuild)
+
+	pkgVersToBuild, err := schedulerutils.CalculatePackagesToBuild(packagesNamesToBuild, packagesNamesToRebuild, *imageConfig, *baseDirPath)
+	if err != nil {
+		logger.Log.Panicf("Error while calculating to build. Error: %s", err)
+	}
+
+	removeExistingGoalNode(scrubbedGraph, allGoalNodeName)
+	_, err = scrubbedGraph.AddGoalNode(allGoalNodeName, pkgVersToBuild, true)
+	if err != nil {
+		logger.Log.Panicf("Unable to set Goal node for packages to build. error: %s", err)
+	}
+
+	err = replaceRunNodesWithPrebuiltNodes(scrubbedGraph, pkgVersToBuild)
+	if err != nil {
+		logger.Log.Panicf("Failed to replace run nodes with preBuilt nodes. Error: %s", err)
+	}
+
+	logger.Log.Infof("Nodes before subgraphing: %d", len(scrubbedGraph.AllNodes()))
+
+	optimizedGraph, err := subGraphPkgsToBuild(scrubbedGraph, pkgVersToBuild)
```

Comment:
What is the benefit of pulling the sub graph here? The build scheduler will attempt to do this at each step (assuming there are no dangling implicit dependencies) and I'm not sure I like throwing away data we might want to use later.